src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs

```diff
@@ -108,6 +111,142 @@
+
+            if (genericName == null)
+            {
+                // When we can't get a generic name, use the offset and the ID.
+                // It is not ideal, but at least it is non-ambiguous.
+                if (timeZoneId.Equals(UtcId, StringComparison.OrdinalIgnoreCase))
+                    displayName = $"(UTC) {timeZoneId}";
+                else
+                    displayName = $"{baseOffsetText} {timeZoneId}";
+                return;
+            }
+
+            // Get the generic location name.
+            string? genericLocationName = null;
+            GetDisplayName(timeZoneId, Interop.Globalization.TimeZoneDisplayNameType.GenericLocation, uiCulture.Name, standardName, ref genericLocationName);
+
+            // Some edge cases only apply when the offset is +00:00.
+            if (baseUtcOffset == TimeSpan.Zero)
+            {
+                // GMT and its aliases will just use the equivalent of "Greenwich Mean Time".
+                string? gmtLocationName = null;
+                GetDisplayName(GmtId, Interop.Globalization.TimeZoneDisplayNameType.GenericLocation, uiCulture.Name, null, ref gmtLocationName);
+                if (genericLocationName == gmtLocationName)
+                {
+                    displayName = $"{baseOffsetText} {genericName}";
+                    return;
+                }
+
+                // Other zones with a zero offset and the equivalent of "Greenwich Mean Time" should only use the location name.
+                // For example, prefer "Iceland Time" over "Greenwich Mean Time (Reykjavik)".
+                string? gmtGenericName = null;
+                GetDisplayName(GmtId, Interop.Globalization.TimeZoneDisplayNameType.Generic, uiCulture.Name, null, ref gmtGenericName);
+                if (genericName == gmtGenericName)
+                {
+                    displayName = $"{baseOffsetText} {genericLocationName}";
+                    return;
+                }
+            }
+
+            if (genericLocationName == genericName)
+            {
+                // When the location name is the same as the generic name,
+                // then it is generally good enough to show by itself.
+
+                // *** Example (en-US) ***
+                // id                   = "America/Havana"
+                // baseOffsetText       = "(UTC-05:00)"
+                // standardName         = "Cuba Standard Time"
+                // genericName          = "Cuba Time"
+                // genericLocationName  = "Cuba Time"
+                // exemplarCityName     = "Havana"
+                // displayName          = "(UTC-05:00) Cuba Time"
+
+                displayName = $"{baseOffsetText} {genericLocationName}";
+                return;
+            }
+
+            // Also use location names in some special cases.  (See the list at the top of this file.)
+            foreach (string id in s_zonesThatUseLocationName)
```

Comment:
>s_zonesThatUseLocationName [](start = 34, length = 26)

instead of defining this as array of strings, define all of them in one string. and then call Indexof on that string.